content/automate/ManageIQ/Cloud/Orchestration/Provisioning/StateMachines/Methods.class/__methods__/check_provisioned.rb

```diff
@@ -47,10 +47,15 @@ def prepare_and_call_refresh_provider(service)
 
               def refresh_provider(service)
                 provider = service.orchestration_manager
+                stack    = service.orchestration_stack
 
-                @handle.log("info", "Refreshing provider #{provider.name}")
+                @handle.log("info", "Refreshing provider '#{provider.name}' and stack '#{stack.name}'")
                 @handle.set_state_var('provider_last_refresh', provider.last_refresh_date.to_i)
-                provider.refresh
+                refreshed_stack = @handle.vmdb(:orchestration_stack).find_by(:name    => stack.name,
+                                                                             :ems_ref => stack.ems_ref)
```

Comment:
@bzwei Can you review these changes?  I do not think we are always going to have a `stack` when this is called so the references to `stack.name` and `stack.ems_ref` will raise errors.

Reply:
@Ladas  Here the `stack` object was inserted by our provisioning service. It does not have all the relationships associated, thus requires a refresh. Chances are the object may be deleted by another full refresh. 

My questions are
1. Will a target refresh work if the stack object was not a result of refresh but inserted by our provisioning logic? I guess the answer is yes.
2. If the stack object happened to be deleted by an earlier full refresh, how can you start a target refresh? I assume you will need a full refresh in this case.
3. A full/target refresh may not be needed at all if event driven refresh is already engaged. We simply need to wait until the stack object appears in vmdb with an expected status.

Reply:
1. so right now this require OrchestrationStack record to exist. I could prepare it as a class method, since all we need for refresh is the ems_ref

2. right, so better would be not to require the record. We can still do targeted refresh, which will create the stack if it exists in the provider.

3. Right, for AWS, we catch the event and refresh stack, but it might be quicker to do a periodic targeted refresh, since the Stack state change event might be slow.

Reply:
ok, let me change it to using OrchestrationStack class method, that way it should work even if we do not receive any event